wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java

```diff
@@ -453,13 +449,19 @@ public void cancel(Command... commands) {
       if (!isScheduled(command)) {
         continue;
       }
-
-      m_scheduledCommands.remove(command);
-      m_requirements.keySet().removeAll(command.getRequirements());
+      // Prevents infinite recursion if cancel() is called from end() by checking if the command is
+      // already being cancelled
+      if (command == m_currentCancelingCommand) {
+        continue;
+      }
+      m_currentCancelingCommand = command;
```

Comment:
Is this something people have run into? Generally commands shouldn't be scheduling/canceling themselves.

Reply:
There's a test where a command calls `cancel` from `end`. `SchedulingRecursionTest` actually has a whole suite of tests where commands call `cancel` or `schedule` from lifecycle methods. Commands really shouldn't be scheduling/canceling themselves, but we seem to put in guard rails just in case it happens.

Reply:
That's a nice touch.  I note that an infinite loop is still possible, e.g.:

Case 1:
* A.end() calls B.cancel()
* B.end() calls A.cancel()

Case 2:
* A.end() calls B.cancel() then A.cancel()

Case 3:
* A.end() calls B.schedule() then A.cancel() where B uses a subsystem with a default command.

Reply:
> Is this something people have run into? Generally commands shouldn't be scheduling/canceling themselves.

See #4529

Reply:
I'm trying to think of a way to defer cancels, but I'm not sure where I would put the commands; I can't modify `m_toCancel` because I'll get a `ConcurrentModificationException`, and since there could be an infinite number of commands calling cancel on the next (think Case 1 from above. but with something like 10 commands), one list isn't enough (and modifying a list will again cause a CME).

Reply:
Maybe if you switch `m_toCancel` to some flavour of `Queue` and `poll` until empty instead of iterating.  Probably need an outer loop to cycle until both pending queues are empty.

Reply:
What should I do for C++? It has a queue I can use, but we always seem to have our own version of containers for performance...

Reply:
I don't work on the C++ much, but I would think `std::queue` would be fine.  `SmallVector` would give you LIFO, but I think FIFO would fit the intention better.